nil/internal/execution/state_accessor.go

```diff
@@ -14,814 +11,241 @@ import (
 	lru "github.com/hashicorp/golang-lru/v2"
 )
 
-type fieldAccessor[T any] func() T
-
-func notInitialized[T any](name string) fieldAccessor[T] {
-	return func() T { panic(fmt.Sprintf("field not initialized : `%s`", name)) }
-}
-
-func initWith[T any](val T) fieldAccessor[T] {
-	return func() T { return val }
-}
-
 /*
-supposed usage is
+StateAccessor supposed usage:
 
-data, err := accessor.Access(tx, shardId).GetBlock().ByHash(hash)
-block := data.Block
+// Share accessor between goroutines, it is thread-safe.
+accessor := execution.NewStateAccessor(1000, 2000)
 
-data, err := accessor.Access(tx, shardId).GetBlock().ByIndex(index)
-block := data.Block
+rawHeader, err := accessor.Access(tx, shardId).GetRawBlockHeaderByHash(hash)
 
-data, err := accessor.Access(tx, shardId).GetBlock().WithInTransactions().ByIndex(index)
-block, txns := data.Block, data.InTransactions
+data, err := accessor.Access(tx, shardId).GetFullBlockByNumber(index)
+header, txns := data.Block, data.InTransactions
 ...
 */
+
 type StateAccessor struct {
-	cache    *accessorCache
-	rawCache *rawAccessorCache
+	txnCache *txnCache
+
+	blockByHash       *BlockByHashAccessor
+	blockHashByNumber *BlockHashByNumberAccessor
 }
 
 func NewStateAccessor(blockLRUSize, txnLRUSize int) *StateAccessor {
```

Comment:
btw if you wanna improve cache it may make sense to implement slightly different algo than simple LRU

https://en.wikipedia.org/wiki/Adaptive_replacement_cache - this one seems to be quite efficient
https://github.com/alexanderGugel/arc - and here's some implementation in go